src/ostree/ot-admin-builtin-finalize-staged.c

```diff
@@ -51,10 +52,58 @@ ot_admin_builtin_finalize_staged (int argc, char **argv, OstreeCommandInvocation
   g_autoptr(GOptionContext) context = g_option_context_new ("");
   g_autoptr(OstreeSysroot) sysroot = NULL;
   if (!ostree_admin_option_context_parse (context, options, &argc, &argv,
-                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER,
+                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER | OSTREE_ADMIN_BUILTIN_FLAG_UNLOCKED,
                                           invocation, &sysroot, cancellable, error))
     return FALSE;
 
+  /* In case they're automounts, open /boot and the booted deployment on
+   * /sysroot so that the automounts don't expire. If they do expire and either
+   * gets unmounted, the service will be stopped.
+   */
+  int sysroot_fd = ostree_sysroot_get_fd (sysroot);
+  glnx_autofd int boot_fd = -1;
+  g_autoptr(OstreeDeployment) deployment = NULL;
+  g_autofree char *deployment_path = NULL;
+  glnx_autofd int deployment_fd = -1;
+  g_debug ("Opening /boot directory");
+  if (!glnx_opendirat (sysroot_fd, "boot", TRUE, &boot_fd, error))
+    return FALSE;
+  deployment = ostree_sysroot_require_booted_deployment (sysroot, error);
+  if (deployment == NULL)
+    return FALSE;
+  deployment_path = ostree_sysroot_get_deployment_dirpath (sysroot, deployment);
+  g_debug ("Opening booted deployment /%s", deployment_path);
+  if (!glnx_opendirat (sysroot_fd, deployment_path, TRUE, &deployment_fd, error))
+    return FALSE;
+
+  /* We want to wait until the system is shut down to actually finalize, so
+   * block on SIGTERM under the assumption that it will be received when
+   * systemd stops the unit.
+   */
+  sigset_t set;
+  int ret, sig;
+  sigemptyset (&set);
+  sigaddset (&set, SIGTERM);
+  ret = sigprocmask (SIG_BLOCK, &set, NULL);
+  if (ret != 0)
+    {
+      errno = ret;
+      return glnx_throw_errno_prefix (error, "sigprogmask");
+    }
+  g_debug ("Waiting for SIGTERM");
+  ret = sigwait (&set, &sig);
```

Comment:
I'm definitely OK with this as is, but we might consider using e.g. `g_unix_signal_add` and blocking in an event loop instead.  That would allow us to generalize in the future to e.g. also using inotify to watch staged deployments.

Reply:
Good call. I had been thinking the same while pondering whether to watch `/run/ostree/staged-deployment`, so might as well just do it now.